spec/formatting.md

```diff
@@ -12,7 +12,7 @@ their handling during formatting is specified here as well.
 
 Formatting of a _message_ is defined by the following operations:
 
-- **_Expression Resolution_** determines the value of an _expression_,
+- **_Expression and Markup Resolution_** determines the value of an _expression_ or _markup_,
```

Comment:
I think we should make *markup resolution* a separate bullet in this list. Resolving markup is a very different operations from resolving expressions due to markup's not calling any functions, due to the default ignorable property in `formatToString`, and due to the fact that it can only happen inside patterns.

In fact, I'm not even sure if we should call it *resolution*. I guess we still resolve options, but crucially, in its current design, markup is encoded in the data model rather directly.

Reply:
> I think we should make _markup resolution_ a separate bullet in this list. Resolving markup is a very different operations from resolving expressions due to markup's not calling any functions, due to the default ignorable property in `formatToString`, and due to the fact that it can only happen inside patterns.

I would prefer to keep these together, because the similarities are much greater than the differences.
- Markup resolution does not call external functions, but its results may be defined in the same way as for expressions. We also have expressions such as literals and variables which do not necessarily call any functions.
- Markup resolution uses the same option resolution as functions.
- The behaviour of markup during string formatting is a concern during _formatting_, but not _resolution_.
- Markup resolution only being used for pattern placeholders does not change how it is done.

> In fact, I'm not even sure if we should call it _resolution_. I guess we still resolve options, but crucially, in its current design, markup is encoded in the data model rather directly.

Could you clarify what you mean by "data model" here? The only data model we define in detail is for the message before formatting, from which we need to select, resolve, and format a pattern during formatting. We left out formatted parts, so the output structure is not well defined.

Reply:
My first reaction to the draft was like @stasm's: it's okay to be somewhat repetitious in a spec. Keeping a wall between the two similar-but-unlike concepts might be useful. 

I didn't make that comment in the end, because there is a lot of commonality and the resolution into different buckets does happen later.

> We also have expressions such as literals and variables which do not necessarily call any functions.

Are you sure? If there is a placeholder `{|zebra|}`, I'm pretty sure that invokes the `:string` formatting function. If there is a placeholder `{$foo}`, I'd allow the implementation to introspect `foo`'s type, but I would expect *some* function to be called in the end. We don't want it to be an error to write messages like `Hello {$user}`...

Reply:
> If there is a placeholder `{|zebra|}`, I'm pretty sure that invokes the `:string` formatting function. If there is a placeholder `{$foo}`, I'd allow the implementation to introspect `foo`'s type, but I would expect _some_ function to be called in the end.

I agree that this is a reasonable thing for an implementation to do, but I don't think we should expect it to _always_ happen. There are also reasonable further questions, such as: What formatter is called for `{|zebra|}` if an implementation allows for overriding default formatters, and `:string` is so overridden? Is it the core `:string` formatter, or the user-provided `:string` formatter that handles the zebra?

I think that's a question we ought to _not_ resolve in this spec, leaving it instead to implementations.

Reply:
> I think that's a question we ought to not resolve in this spec, leaving it instead to implementations.

I agree. Hmm... what concrete change are we proposing with this thread?